tests/kernel/timer/timer_behavior/src/tick_timer_train.c

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/time_units.h>
+#include <zephyr/tc_util.h>
+#include <zephyr/ztest.h>
+
+#define TIMERS 1
+#define TEST_SECONDS 10
+#define MAX_CALLBACKS (CONFIG_SYS_CLOCK_TICKS_PER_SEC*TEST_SECONDS)
+
+struct timer_wrapper {
+	uint32_t callbacks;
+	struct k_timer tm;
+	uint32_t last_isr;
+	int64_t last_scheduled;
+	uint32_t max_delta;
+};
+
+K_SEM_DEFINE(timers_sem, 0, K_SEM_MAX_LIMIT);
+
+struct timer_wrapper timers[TIMERS];
+
+void tm_fn(struct k_timer *tm)
+{
+	struct timer_wrapper *tm_wrap =
+		CONTAINER_OF(tm, struct timer_wrapper, tm);
+	uint32_t now = k_cycle_get_32();
+
+	if (tm_wrap->last_isr != 0) {
+		uint32_t delta = now - tm_wrap->last_isr;
+
+		tm_wrap->max_delta = delta > tm_wrap->max_delta ? delta : tm_wrap->max_delta;
+	}
+	tm_wrap->last_isr = now;
+	tm_wrap->callbacks++;
+	if (tm_wrap->callbacks >= MAX_CALLBACKS) {
+		k_timer_stop(tm);
+		k_sem_give(&timers_sem);
+	} else {
+		int64_t next = tm_wrap->last_scheduled + TIMERS;
+
+		tm_wrap->last_scheduled = next;
+		k_timer_start(tm, K_TIMEOUT_ABS_TICKS(next), K_NO_WAIT);
```

Comment:
Given that the provenance of the test is to exercise timer agility, I don't know that you want to be doing this much work in the callback.  Why not just use a periodic timer with a 1 tick period?  Basically: there are faster ways to do this.  Since this is a performance test, we should be exercising the fast paths that performance-sensitive apps will take.

Reply:
I tried doing that, but it always took twice as long as I expected, I think because of the + 1 in timer.c for periodically scheduled timers was my thought but I didn't look that hard at it to be honest.